File: backend/models/tool.py
from app import db
from datetime import datetime
import uuid

class Tool(db.Model):
    __tablename__ = 'tools'
    
    id = db.Column(db.Integer, primary_key=True)
    uuid = db.Column(db.String(36), unique=True, nullable=False, default=lambda: str(uuid.uuid4()))
    name = db.Column(db.String(100), nullable=False)
    description = db.Column(db.Text, nullable=True)
    category = db.Column(db.String(50), nullable=True)  # 工具分类
    url = db.Column(db.String(255), nullable=True)  # 工具链接
    icon = db.Column(db.String(255), nullable=True)  # 工具图标
    
    # 状态管理
    is_public = db.Column(db.Boolean, default=False)  # 是否公开
    status = db.Column(db.String(20), default='pending')  # pending, approved, rejected
    
    # 关联用户
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    user = db.relationship('User', backref=db.backref('tools', lazy=True))
    
    # 审核信息
    reviewed_by = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=True)
    reviewed_at = db.Column(db.DateTime, nullable=True)
    review_comment = db.Column(db.Text, nullable=True)
    
    # 时间戳
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    def approve(self, admin_user_id, comment=None):
        """审核通过"""
        self.status = 'approved'
        self.reviewed_by = admin_user_id
        self.reviewed_at = datetime.utcnow()
        self.review_comment = comment
        self.updated_at = datetime.utcnow()
    
    def reject(self, admin_user_id, comment=None):
        """审核拒绝"""
        self.status = 'rejected'
        self.reviewed_by = admin_user_id
        self.reviewed_at = datetime.utcnow()
        self.review_comment = comment
        self.updated_at = datetime.utcnow()
    
    def set_public(self, is_public):
        """设置公开状态"""
        self.is_public = is_public
        if is_public:
            self.status = 'pending'  # 设为公开时需要重新审核
        self.updated_at = datetime.utcnow()
    
    def to_dict(self, include_review_info=False):
        """转换为字典格式"""
        result = {
            'id': self.id,
            'uuid': self.uuid,
            'name': self.name,
            'description': self.description,
            'category': self.category,
            'url': self.url,
            'icon': self.icon,
            'is_public': self.is_public,
            'status': self.status,
            'user_id': self.user_id,
            'user_name': self.user.name if self.user else None,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None
        }
        
        if include_review_info:
            result.update({
                'reviewed_by': self.reviewed_by,
                'reviewed_at': self.reviewed_at.isoformat() if self.reviewed_at else None,
                'review_comment': self.review_comment
            })
        
        return result
    
    def __repr__(self):
        return f'<Tool {self.name}>'

File: backend-nodejs/models/Tool.js
const { DataTypes } = require('sequelize');
const { v4: uuidv4 } = require('uuid');
const { sequelize } = require('../config/database-connection');

const Tool = sequelize.define('Tool', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  uuid: {
    type: DataTypes.STRING(36),
    allowNull: false,
    unique: true,
    defaultValue: () => uuidv4()
  },
  name: {
    type: DataTypes.STRING(100),
    allowNull: false,
    validate: {
      notEmpty: true,
      len: [1, 100]
    }
  },
  description: {
    type: DataTypes.TEXT,
    allowNull: true
  },
  category: {
    type: DataTypes.STRING(50),
    allowNull: false,
    validate: {
      notEmpty: true
    }
  },
  url: {
    type: DataTypes.STRING(500),
    allowNull: false,
    validate: {
      isUrl: true,
      notEmpty: true
    }
  },
  icon: {
    type: DataTypes.STRING(500),
    allowNull: true,
    validate: {
      isUrl: {
        msg: 'Icon must be a valid URL'
      }
    }
  },
  is_public: {
    type: DataTypes.BOOLEAN,
    allowNull: false,
    defaultValue: false
  },
  status: {
    type: DataTypes.ENUM('pending', 'approved', 'rejected'),
    allowNull: false,
    defaultValue: 'pending'
  },
  user_id: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: 'users',
      key: 'id'
    }
  },
  approved_by: {
    type: DataTypes.INTEGER,
    allowNull: true,
    references: {
      model: 'users',
      key: 'id'
    }
  },
  approved_at: {
    type: DataTypes.DATE,
    allowNull: true
  },
  created_at: {
    type: DataTypes.DATE,
    allowNull: false,
    defaultValue: DataTypes.NOW
  },
  updated_at: {
    type: DataTypes.DATE,
    allowNull: false,
    defaultValue: DataTypes.NOW
  }
}, {
  tableName: 'tools',
  timestamps: true,
  createdAt: 'created_at',
  updatedAt: 'updated_at',
  hooks: {
    beforeUpdate: (tool) => {
      tool.updated_at = new Date();
    }
  }
});

// 实例方法：批准工具
Tool.prototype.approve = function(approvedBy) {
  this.status = 'approved';
  this.approved_by = approvedBy;
  this.approved_at = new Date();
  this.is_public = true;
};

// 实例方法：拒绝工具
Tool.prototype.reject = function(approvedBy) {
  this.status = 'rejected';
  this.approved_by = approvedBy;
  this.approved_at = new Date();
  this.is_public = false;
};

// 实例方法：设置公开状态
Tool.prototype.setPublic = function(isPublic) {
  this.is_public = Boolean(isPublic);
};

// 实例方法：转换为字典
Tool.prototype.toDict = function(includeUser = false) {
  const toolDict = {
    id: this.id,
    uuid: this.uuid,
    name: this.name,
    description: this.description,
    category: this.category,
    url: this.url,
    icon: this.icon,
    is_public: this.is_public,
    status: this.status,
    user_id: this.user_id,
    approved_by: this.approved_by,
    approved_at: this.approved_at,
    created_at: this.created_at,
    updated_at: this.updated_at
  };

  if (includeUser && this.User) {
    toolDict.user = this.User.toDict();
  }

  if (includeUser && this.ApprovedBy) {
    toolDict.approved_by_user = this.ApprovedBy.toDict();
  }

  return toolDict;
};

// 类方法：获取公开工具
Tool.getPublicTools = async function(options = {}) {
  const { category, limit, offset } = options;
  const whereClause = { 
    is_public: true, 
    status: 'approved' 
  };

  if (category) {
    whereClause.category = category;
  }

  return await this.findAndCountAll({
    where: whereClause,
    limit: limit || 20,
    offset: offset || 0,
    order: [['created_at', 'DESC']],
    include: [{
      model: require('./User'),
      as: 'User',
      attributes: ['id', 'name', 'email', 'avatar']
    }]
  });
};

// 类方法：获取待审核工具
Tool.getPendingTools = async function(options = {}) {
  const { limit, offset } = options;

  return await this.findAndCountAll({
    where: { status: 'pending' },
    limit: limit || 20,
    offset: offset || 0,
    order: [['created_at', 'ASC']],
    include: [{
      model: require('./User'),
      as: 'User',
      attributes: ['id', 'name', 'email', 'avatar']
    }]
  });
};

module.exports = Tool;